fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -1971,6 +2121,216 @@ struct ShardedRocksDBKeyValueStore : IKeyValueStore {
+
+			rocksdb::PinnableSlice value;
+			rocksdb::ReadOptions readOptions = getReadOptions();
+			s = a.shardManager->getDb()->Get(
+			    readOptions, a.shardManager->getSpecialKeysShard()->cf, toSlice(persistVersion), &value);
+
+			if (!s.ok() && !s.IsNotFound()) {
+				logRocksDBError(s, "Checkpoint");
+				a.reply.sendError(statusToError(s));
+				return;
+			}
+
+			const Version version = s.IsNotFound()
+			                            ? latestVersion
+			                            : BinaryReader::fromStringRef<Version>(toStringRef(value), Unversioned());
+
+			TraceEvent(SevDebug, "ShardedRocksDBServeCheckpointVersion", logId)
+			    .detail("CheckpointVersion", a.request.version)
+			    .detail("PersistVersion", version);
+			ASSERT(a.request.version == version || a.request.version == latestVersion);
+
+			CheckpointMetaData res(ps->getAllRanges(), version, a.request.format, a.request.checkpointID);
+			const std::string& checkpointDir = abspath(a.request.checkpointDir);
+
+			if (a.request.format == DataMoveRocksCF) {
+				rocksdb::ExportImportFilesMetaData* pMetadata;
+				platform::eraseDirectoryRecursive(checkpointDir);
+				s = checkpoint->ExportColumnFamily(ps->cf, checkpointDir, &pMetadata);
+				if (!s.ok()) {
+					logRocksDBError(s, "ExportColumnFamily");
+					a.reply.sendError(statusToError(s));
+					return;
+				}
+
+				populateMetaData(&res, *pMetadata);
+				delete pMetadata;
+				TraceEvent(SevInfo, "ShardedRocksDBServeCheckpointSuccess", logId)
+				    .detail("CheckpointMetaData", res.toString())
+				    .detail("RocksDBCF", getRocksCF(res).toString());
+			} else {
+				if (checkpoint != nullptr) {
+					delete checkpoint;
+				}
+				a.reply.sendError(not_implemented());
+				return;
+			}
+
+			if (checkpoint != nullptr) {
+				delete checkpoint;
+			}
+			res.setState(CheckpointMetaData::Complete);
+			a.reply.send(res);
+		}
+
+		struct RestoreAction : TypedAction<Writer, RestoreAction> {
+			RestoreAction(ShardManager* shardManager,
+			              const std::string& path,
+			              const std::string& shardId,
+			              const std::vector<KeyRange>& ranges,
+			              const std::vector<CheckpointMetaData>& checkpoints)
+			  : shardManager(shardManager), path(path), shardId(shardId), ranges(ranges), checkpoints(checkpoints) {}
+
+			double getTimeEstimate() const override { return SERVER_KNOBS->COMMIT_TIME_ESTIMATE; }
+
+			ShardManager* shardManager;
+			const std::string path;
+			const std::string shardId;
+			std::vector<KeyRange> ranges;
+			std::vector<CheckpointMetaData> checkpoints;
+			ThreadReturnPromise<Void> done;
+		};
+
+		void action(RestoreAction& a) {
+			TraceEvent(SevInfo, "ShardedRocksDBRestoreBegin", logId)
+			    .detail("Path", a.path)
+			    .detail("Checkpoints", describe(a.checkpoints));
+
+			ASSERT(!a.checkpoints.empty());
+
+			PhysicalShard* ps = nullptr;
+			for (const KeyRange& range : a.ranges) {
+				ps = a.shardManager->addRange(range, a.shardId);
```

Comment:
`addRange` changes internal shard mapping. It should only be called on main thread.

Reply:
Nice catch! Moved the block to the main thread.